L10 unittest

Tasenga opened the pull request:
Added unit tests.
Updated readme and requirements.
Small functions fix was done (because unit tests revealed problems).
Git.hooks pre-push was configured to check unit test results.

Commit:
31 unittest were created.

fixed minor mistakes in common modules

Commit:
updated requirements

Commit:
added check unit test before pushing

Commit:
import Path

aliaksandr-tsukanau commented on FP_analysis_of_reviews_from_Amazon/Task_9_1.py:
```diff
@@ -20,25 +21,26 @@ def bad_comment(data):
-            [review for review in reviews
-             if all(key not in review.keys() for key in required_keys)]
+            [review for review in data
+             if any(key not in review.keys() for key in required_keys)]
```
let's use [generator expressions](https://www.python.org/dev/peps/pep-0289/) inside all `len`, `min`, `sum` etc.

Tasenga replied:
I read documentation and understood idea, it seems very effectively. But I don't know how to use this in my code. For example, I have 'best_comment' function:
```
def best_comment(data):
    return max(data, key=lambda review: review["helpful"][0],
                      default={"asin": None, "helpful": [None, None], "reviewText": None})
```
I expect that I will receive something like that:
{
      asin: ...,
      reviewerID:....
      helpful: [found max value, ....],
      textreview: ...,
      ...
}
or default value, if max value isn't exist.
But, if I use generator expression in this function:
```
def best_comment(data):
    return max(review["helpful"][0] for review in data)
```
I will receive integer, which equal max value. Such variant isn't suit for me.
How I understood, this variant isn't generator expression and equal the first variant:
```
def best_comment(data):
    return max((review for review in data), key=lambda review: review["helpful"][0],
                      default={"asin": None, "helpful": [None, None], "reviewText": None})
```
Just in case, I redid everything like third variant, except len, because I received exception 'generation object doesn't have attribute len'.

sleshJdev replied:
Lemmy jump in. @Tasenga in your last example:
```
def best_comment(data):
    return max((review for review in data), key=lambda review: review["helpful"][0],
                      default={"asin": None, "helpful": [None, None], "reviewText": None})
```
`(review for review in data)` it's actually a generator expression, but you applied it in the not best way. The `data` will actually be iterated twice: by generator and by max func, which is overhead in this particular case.  

Let's better consider this example:
```
def count_unanalyzed(*required_keys):
        return len(
            [review for review in data
             if any(key not in review.keys() for key in required_keys)]
        )
```

How generators can be applied here?

Tasenga replied:
I returned previous implementation of best_comment.
I replaced len with sum everywhere if it was possible, such an example:
```
    def count_unanalyzed(*required_keys):
        return sum(
            1 for review in data
             if any(key not in review.keys() for key in required_keys)
        )
```

aliaksandr-tsukanau replied:
I'm sorry I was not very precise. This code:
```
return max(data, key=lambda review: review["helpful"][0],
                      default={"asin": None, "helpful": [None, None], "reviewText": None})
```
is absolutely fine. The idea was to avoid list comprehensions and use generator expressions instead of them where possible. Not literally *all* len, min and sum needed to be editied, sorry for that phrase

aliaksandr-tsukanau replied:
`sum(1 for ...)` is a good idea, this'll save memory

aliaksandr-tsukanau replied:
@Tasenga I cannot see `sum(1 ...` being used in your code, please update

Tasenga replied:
I'm sorry, updated. It is because branch has been renamed

aliaksandr-tsukanau commented on FP_analysis_of_reviews_from_Amazon/nearest_review.py:
```diff
@@ -25,85 +25,88 @@ def prepare_data(data):
-        for i in range(1, len(data))
-    )
+    deltatime = []
+    for i in range(1, len(data)):
```
I would iterate over `zip(data[:-1], data[1:])` to avoid indices but it's a matter of personal choice

Tasenga replied:
I don't know how I prefer to do, so I tried with zip. All managed.

aliaksandr-tsukanau commented on FP_analysis_of_reviews_from_Amazon/nearest_review.py:
```diff
@@ -25,85 +25,88 @@ def prepare_data(data):
 
 
-def get_potential_bots(data):
+def get_potential_bots(data=[]):
```
[Mutable default argument](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments)

Tasenga replied:
I see. Fixed.

aliaksandr-tsukanau commented on FP_analysis_of_reviews_from_Amazon/test_nearest_review.py:
```diff
@@ -0,0 +1,267 @@
+        self.assertEqual(result["1"][1].text, self.data[0]["reviewText"])
+        self.assertEqual(result["1"][2].text, self.data[1]["reviewText"])
+        self.assertTrue(
+            i.text is str or i.text is None for v in result.values() for i in v
```
`i.text is str` - are you trying to check the type? Then please use `isinstance`. `is` is used to compare two objects by reference, not the type. Check elsewhere also

Tasenga replied:
I tried to check each element in the second position from an object from the list that it is an instance of a string or non-type object. Would it be better to split it into multiple checks for each element using assertIsInstance?

aliaksandr-tsukanau replied:
It is fine to do either multiple `assertIsInstance` or this way:
```python
self.assertTrue(all(isinstance(i.text, str) or i.text is None for ...))
```
Your tests pass because the code is incorrect twice. `i.text` never *is* `str` (but it `isinstance` of `str`). Inside `assertTrue` you construct a generator which produces `False` for each `i`. But since generator is not empty it is considered a `True` value and the test passes. You can read more about [`__bool__`](https://docs.python.org/3/reference/datamodel.html#object.__bool__) method to understand

aliaksandr-tsukanau replied:
please fix or ask me if you have more questions on this

Tasenga replied:
Done

Commit:
used generator expression;

fixed function with mutable default argument;
used zip(data[:-1], data[1:]) for iteration in 'count_deltatime' function.

sleshJdev commented:
Please, resolve conflicts

Tasenga committed:
Merge branch 'master' into L11_unittest

Tasenga commented:
> 
> 
> Please, resolve conflicts

resolved

sleshJdev commented on FP_analysis_of_reviews_from_Amazon/nearest_review.py:
```diff
@@ -25,85 +25,90 @@ def prepare_data(data):
-        for i in range(1, len(data))
-    )
+    deltatime = []
+    for el1, el2 in zip(data[:-1], data[1:]):
```
It seems to be a good candidate to try generators+list comprehensive

Tasenga replied:
I couldn't raise exception in this case.
```
def count_deltatime(data):
    return min(
        (el2.time - el1.time, el2.text, el1.text)
        if el2.time >= el1.time
        else raise Exception
        for el1, el2
        in zip(data[:-1], data[1:])
    )
```
I couldn't find this case in documentation. Is it really impossible to do?

aliaksandr-tsukanau replied:
It is indeed impossible since it would have made code less readable if it was allowed. for-loop is excellent here IMO, let's keep things as is

Tasenga replied:
Saved without changes

Commit:
method len was replaced by method sum with generator expression

Commit:
Merge branch 'L11_unittest' of https://github.com/Tasenga/ITechArt_Android_App_Review into L10_unittest

sleshJdev commented on FP_analysis_of_reviews_from_Amazon/test_common_functions.py:
```diff
@@ -0,0 +1,84 @@
+            get_apps_scores(self.data), {"1": [5.0], "2": [5.0, 4.0], "3": [4.0]}
+        )
+    def test_get_apps_scores_without_overall(self):
+        self.data[0].pop("overall")
```
Mmm, can we just create a new dataset? Don't put extra logic into the tests if that's possible

Tasenga replied:
I thought that it is important to show differences with initial data to create new subtests. But of cause I fixed and created new dataset everywhere. Done.

sleshJdev commented on FP_analysis_of_reviews_from_Amazon/test_common_functions.py:
```diff
@@ -0,0 +1,84 @@
+        self.assertEqual(
+            get_apps_scores(self.data), {"1": [1.0, 5.0], "2": [5.0, 4.0], "3": [4.0]}
+        )
+    def test_get_apps_scores_with_overall_exc(self):
```
What is _exc?

Tasenga replied:
renamed this function everywhere. I mean test_get_dict_of_apps_with_score_without_one_overall

sleshJdev commented on FP_analysis_of_reviews_from_Amazon/test_common_functions.py:
```diff
@@ -0,0 +1,84 @@
+    def test_get_dict_of_apps_with_score(self):
+        result = get_dict_of_apps_with_score([self.data1, self.data2])
+        self.assertCountEqual(result.keys(), ["1", "2", "3"])
+        self.assertEqual(result["1"].asin, "1")
```
We can better. Try to implement [__eq__](https://docs.python.org/3/reference/datamodel.html#object.__eq__) method and replace this per property assert with one equal one:
```
        self.assertEqual(
            get_dict_of_apps_with_score([
                self.data1, self.data2]),
            {"1": AppScore("1", 6, 2),
             "2": AppScore("2", 9, 2),
             "3": AppScore("3", 4, 1)})
```

Make sure to do the same for the rest tests, where it make sense.

sleshJdev replied:
As @alex-tsukanov mentioned just below, you can use dataclasses to handle classes that contain only data and not business logic. The dataclass annotation has also parameter **eq** which can be used to generate __eq__ method automatically. Please, let's do so.

Tasenga replied:
I used dataclass everywhere with using this check in test:
```
        self.assertEqual(
            get_dict_of_apps_with_score([
                self.data1, self.data2]),
            {"1": AppScore("1", 6, 2),
             "2": AppScore("2", 9, 2),
             "3": AppScore("3", 4, 1)})
```
It works.

aliaksandr-tsukanau commented:
@Tasenga you may want to read about [dataclass decorator](https://docs.python.org/3/library/dataclasses.html). I can see a lot of simple data classes in your code, they can be written simpler. (not necessary to update this time, just for your information)

Commit:
fixed test with using assertequal for result and decorator from dataclass library.

exclude unnecessary logic

Tasenga commented:
Done

aliaksandr-tsukanau commented on FP_analysis_of_reviews_from_Amazon/test_nearest_review.py:
```diff
@@ -0,0 +1,327 @@
+            "3": [ReviewTimeInfo(200, None)],
+        }
+        analyzed_data, potential_bots, number_of_bot_comments = process_data(data)
+        self.assertTrue(v is ReviewTimeInfo for v in analyzed_data.values())
```
I think you've accidentally missed this one

Tasenga replied:
Yes. I deleted it now

Commit:
deleted testcase 'self.assertTrue(v is ReviewTimeInfo for v in analyzed_data.values())'

aliaksandr-tsukanau commented:
looks good to me. @sleshJdev please approve

sleshJdev reviewed (approved):
LGTM, feel free to merge
